### PR TITLE
commands: improve and simplify locking + lxccontainer: add reboot2() API extension

### PR DESCRIPTION
**Pull request by brauner**
lxccontainer: add reboot2() API extension
This adds reboot2() as a new API extension. This function properly wait until a
reboot succeeded. It takes a timeout argument. When set to > 0 reboot2() will
block until the timeout is reached, if timeout is set to zero reboot2() will
not block, if set to -1 reboot2() will block indefinitly.

The struct state_client gets rename to lxc_state_client since it's more in line
with other declarations. It also gets moved from the lxc_handler to the
lxc_conf struct so that the state clients waiting for reboots don't get
deallocated on reboot since the handler is deallocated on reboot.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
jenkins: test this please

File: src/lxc/commands.c [hunk 1]
```diff
@@ -904,44 +904,35 @@ int lxc_cmd_add_state_client(const char *name, const char *lxcpath,
 	    },
 	};
 
-	/* Lock the whole lxc_cmd_add_state_client_callback() call to ensure
-	*  that lxc_set_state() doesn't cause us to miss a state.
-	 */
-	process_lock();
 	/* Check if already in requested state. */
 	state = lxc_getstate(name, lxcpath);
```
Comment (hallyn) [inline src/lxc/commands.c]: ...  but the state could change between lxc_getstate and later checks, no?

File: src/lxc/start.c [hunk 2]
```diff
@@ -341,14 +341,10 @@ static int lxc_serve_state_clients(const char *name,
 	struct state_client *client;
 	struct lxc_msg msg = {.type = lxc_msg_state, .value = state};
 
-	process_lock();
-
-	/* Only set state under process lock held so that we don't cause
-	*  lxc_cmd_add_state_client() to miss a state.
```
Comment (hallyn) [inline src/lxc/start.c]: TBH given the comment here I don't understand why the process_unlock happens before lxc_monitor_send_state().  It seems like the set_state and send_state should happen atomically together.

File: src/lxc/commands.c [hunk 3]
```diff
@@ -914,24 +914,6 @@ int lxc_cmd_add_state_client(const char *name, const char *lxcpath,
 		return state;
 	}
 
-	if ((state == STARTING) && !states[RUNNING] && !states[STOPPING] && !states[STOPPED]) {
```
Comment (hallyn) [inline src/lxc/commands.c]: ...  I'm not quite sure what this code was originally meant to do?

File: src/lxc/start.c [hunk 4]
```diff
@@ -222,6 +225,20 @@ int lxc_check_inherited(struct lxc_conf *conf, bool closeall,
 		    (i < len_fds && fd == fds_to_ignore[i]))
 			continue;
 
+		/* Keep state clients that wait on reboots. */
+		lxc_list_for_each(cur, &conf->state_clients) {
+			struct lxc_state_client *client = cur->elem;
```
Comment (hallyn) [inline src/lxc/start.c]: Seems like this should be a separate helper?

File: src/lxc/conf.h [hunk 5]
```diff
@@ -248,6 +248,11 @@ enum lxchooks {
 
 extern char *lxchook_names[NUM_LXC_HOOKS];
 
+struct lxc_state_client {
```
Comment (hallyn) [inline src/lxc/conf.h]: Can you explain in the commit message for future readers why the lxc_state_client gets moved from handler to conf?

**Comment by brauner** [comment]
On Fri, Nov 24, 2017 at 02:40:21AM +0000, Serge Hallyn wrote:
> hallyn commented on this pull request.
> 
> 
> 
> > @@ -904,44 +904,35 @@ int lxc_cmd_add_state_client(const char *name, const char *lxcpath,
>  	    },
>  	};
>  
> -	/* Lock the whole lxc_cmd_add_state_client_callback() call to ensure
> -	*  that lxc_set_state() doesn't cause us to miss a state.
> -	 */
> -	process_lock();
>  	/* Check if already in requested state. */
>  	state = lxc_getstate(name, lxcpath);
> 
> ...  but the state could change between lxc_getstate and later checks, no?

I think the right thing to do is to remove this check altogether. Instead, we
check the state by looking at handler->state directly when we have taken the
lock on the state client list in lxc_add_state_client(), i.e. in the callback.

> 
> > @@ -341,14 +341,10 @@ static int lxc_serve_state_clients(const char *name,
>  	struct state_client *client;
>  	struct lxc_msg msg = {.type = lxc_msg_state, .value = state};
>  
> -	process_lock();
> -
> -	/* Only set state under process lock held so that we don't cause
> -	*  lxc_cmd_add_state_client() to miss a state.
> 
> TBH given the comment here I don't understand why the process_unlock happens before lxc_monitor_send_state().  It seems like the set_state and send_state should happen atomically together.

This is the legacy lxc-monitord binary which doesn't use the command socket at
all.

> 
> > @@ -914,24 +914,6 @@ int lxc_cmd_add_state_client(const char *name, const char *lxcpath,
>  		return state;
>  	}
>  
> -	if ((state == STARTING) && !states[RUNNING] && !states[STOPPING] && !states[STOPPED]) {
> 
> ...  I'm not quite sure what this code was originally meant to do?
> 
> > @@ -222,6 +225,20 @@ int lxc_check_inherited(struct lxc_conf *conf, bool closeall,
>  		    (i < len_fds && fd == fds_to_ignore[i]))
>  			continue;
>  
> +		/* Keep state clients that wait on reboots. */
> +		lxc_list_for_each(cur, &conf->state_clients) {
> +			struct lxc_state_client *client = cur->elem;
> 
> Seems like this should be a separate helper?

I'm not sure this licenses a separate helper.

File: src/lxc/commands.c (hunk 1 above)
Comment (brauner) [reply src/lxc/commands.c]: I think the right thing to do is to remove this check altogether. Instead, we
check the state by looking at handler->state directly when we have taken the
lock on the state client list in lxc_add_state_client(), i.e. in the callback.

File: src/lxc/conf.h (hunk 5 above)
Comment (brauner) [reply src/lxc/conf.h]: Yeah, good idea.

File: src/lxc/start.c (hunk 2 above)
Comment (brauner) [reply src/lxc/start.c]: This is the legacy lxc-monitord binary which doesn't use the command socket at
all.

File: src/lxc/start.c (hunk 4 above)
Comment (brauner) [reply src/lxc/start.c]: I'm not sure this licenses a separate helper.

**Comment by brauner** [comment]
@hallyn updated.

File: src/lxc/lxccontainer.c [hunk 6]
```diff
@@ -1771,24 +1771,52 @@ static bool lxcapi_create(struct lxc_container *c, const char *t,
...
+		state = lxc_cmd_sock_rcv_state(sock, 0);
+		close(sock);
+		if (state != RUNNING)
+			return false;
+		TRACE("Container rebooted");
+		retv = true;
+	} else {
+		TRACE("Received state \"%s\" instead of expected \"RUNNING\"",
+		      lxc_state2str(ret));
+	}
```
Comment (stgraber) [inline src/lxc/lxccontainer.c]: Hmm, so am I missing something or is this turning reboot() from a non-blocking API call to a blocking API call?

Unlike shutdown() there is no timeout argument here for the caller to indicate that they just want the signal sent but not wait for something to actually happen.

Comment (brauner) [reply src/lxc/lxccontainer.c]: So probably better to add a `reboot2()` that does both jobs. API-wise it'll be pretty nice to finally having a blocking reboot call...

Comment (stgraber) [reply src/lxc/lxccontainer.c]: Yeah, I'd be fine with keeping reboot() be just signal sending and reboot2() be signal sending + wait with a timeout.

File: src/lxc/conf.h (hunk 5 above)
Comment (brauner) [reply src/lxc/conf.h]: done

**Comment by brauner** [comment]
Ready for another review //cc @hallyn.

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
jenkins seems to be messed up

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
jenkins: test this please

**Comment by hallyn** [comment]
Just browsed through the .diff again, lgtm.

**Comment by brauner** [comment]
@hallyn, this should be good to go. :)

**Comment by brauner** [comment]
@stgraber, since @hallyn basically acked this we should land this.

**Pull request merged**